### PR TITLE
TypeScript: Convert `highlights` provider

### LINKED ISSUE #12665: TypeScript: Convert highlights provider

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Testing Instructions

- [x] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

## Checklist

- [ ] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [ ] I have tested this code to the best of my abilities
- [ ] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [ ] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [ ] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [ ] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #12665

### CODE REVIEW

File: packages/story-editor/src/app/quickActions/utils/getResetProperties.js [hunk 1]
```diff
@@ -21,7 +21,7 @@ import { ELEMENT_TYPES } from '@googleforcreators/elements';
 /**
  * Internal dependencies
  */
-import { RESET_PROPERTIES, RESET_DEFAULTS } from '../constants';
+import { RESET_PROPERTIES, RESET_DEFAULTS } from '../../highlights/constants';
```
Comment (swissspidy) [inline packages/story-editor/src/app/quickActions/utils/getResetProperties.js]: These constants should also be moved to the `quickActions` folder and out of `highlights`

File: packages/story-editor/src/app/highlights/styles.ts [hunk 2]
```diff
@@ -18,10 +18,15 @@
  */
 import { rgba } from 'polished';
 import { css, keyframes } from 'styled-components';
-import {
-  theme as dsTheme,
-  themeHelpers,
-} from '@googleforcreators/design-system';
+import type { DefaultTheme } from 'styled-components';
+import { theme, themeHelpers } from '@googleforcreators/design-system';
```
Comment (spacedmonkey) [inline packages/story-editor/src/app/highlights/styles.ts]: @miina @swissspidy These errors are valid but I am not sure how to work around these. 

```
Module '"@googleforcreators/design-system"' has no exported member 'theme'.
```
```
Module '"@googleforcreators/design-system"' has no exported member 'themeHelpers'.
```

Comment (miina) [reply packages/story-editor/src/app/highlights/styles.ts]: One option is doing something like in the `output` package until #12659 gets done:
https://github.com/GoogleForCreators/web-stories-wp/blob/0822033889314d0ffa431e43cfc0860e83e4e2e4/packages/output/src/types.ts#L30

File: packages/story-editor/tsconfig.json [hunk 3]
```diff
@@ -25,6 +25,10 @@
     "src/app/config",
     "src/app/currentUser",
     "src/app/helpCenter",
+    "src/app/helpCenter/**/*",
+    "src/components/library/paneIds.ts",
+    "src/components/sidebar/constants.ts",
+    "src/app/highlights/*",
```
Comment (spacedmonkey) [inline packages/story-editor/tsconfig.json]:
```suggestion
    "src/components/library/paneIds.ts",
    "src/components/sidebar/constants.ts",
    "src/app/highlights",
```

File: packages/story-editor/src/app/highlights/styles.ts (hunk 2 above)
Comment (spacedmonkey) [reply packages/story-editor/src/app/highlights/styles.ts]: I think we should not do this as part of this change. I think we should loop back on this change when theme is converted to typescript later.

Comment (miina) [reply packages/story-editor/src/app/highlights/styles.ts]: Not sure when the theme will be converted though considering the fixit sprint.

File: packages/story-editor/tsconfig.json [hunk 4]
```diff
@@ -25,6 +25,10 @@
     "src/app/config",
     "src/app/currentUser",
     "src/app/helpCenter",
+    "src/components/library/paneIds.ts",
+    "src/components/sidebar/constants.ts",
+    "src/app/highlights/*.ts",
+    "src/app/highlights/provider.tsx",
```
Comment (miina) [inline packages/story-editor/tsconfig.json]: Just `src/app/highlights` should work, no? Or are there some files you're not converting?

File: packages/story-editor/src/app/highlights/states.ts [hunk 5]
```diff
@@ -22,147 +22,111 @@ import { STYLE_PANE_IDS } from '@googleforcreators/design-system';
...
-  // LIBRARY
-  MEDIA: 'MEDIA',
-  MEDIA3P: 'MEDIA3P',
-  TEXT_SET: 'TEXT',
-  PAGE_TEMPLATES: 'PAGE_TEMPLATES',
-
-  // DOCUMENT
-  PAGE_BACKGROUND_AUDIO: 'PAGE_BACKGROUND_AUDIO',
-};
+import { HighlightType } from '../../types/highlightsProvider';
```
Comment (miina) [inline packages/story-editor/src/app/highlights/states.ts]: Hm, the enum of keys should probably still stay here -- since it has actual values and is not just a type.
I suppose that's one of the reasons why enums are not great -- that they change the code instead of just adding types. cc @barklund for thoughts here as the new enum-master.

File: packages/story-editor/src/components/panels/design/animation/animation.js [hunk 6]
```diff
@@ -263,7 +264,7 @@ function AnimationPanel({
             disabledTypeOptionsMap={disabledTypeOptionsMap}
             direction={getEffectDirection(updatedAnimations[0])}
             selectedEffectType={updatedAnimations[0]?.type}
-            selectButtonStylesOverride={highlight?.focus && styles.OUTLINE}
+            selectButtonStylesOverride={highlight?.focus && OUTLINE}
```
Comment (miina) [inline packages/story-editor/src/components/panels/design/animation/animation.js]: Curious why this change? Moving styles separately.

File: packages/story-editor/src/components/canvas/karma/quickActions.karma.js [hunk 7]
```diff
@@ -23,7 +23,7 @@ import { TEXT_ELEMENT_DEFAULT_FONT } from '@googleforcreators/elements';
  * Internal dependencies
  */
 import { useStory } from '../../../app';
-import { ACTIONS } from '../../../app/highlights';
+import { ACTIONS } from '../../../app/quickActions';
```
Comment (miina) [inline packages/story-editor/src/components/canvas/karma/quickActions.karma.js:26]: Here, too -- curious why the change -- moving some files into `quickActions` folder instead.

File: packages/story-editor/src/app/highlights/useHighlights.ts [hunk 8]
```diff
@@ -20,9 +20,14 @@ import { useContextSelector, identity } from '@googleforcreators/react';
...
+import type { HighlightsState } from '../../types/highlightsProvider';
 import Context from './context';
 
-function useHighlights(selector) {
+function useHighlights(): HighlightsState;
+function useHighlights<T>(selector: (state: HighlightsState) => T): T;
+function useHighlights<T>(
+  selector: (state: HighlightsState) => T | HighlightsState = identity
+) {
   return useContextSelector(Context, selector ?? identity);
```
Comment (miina) [inline packages/story-editor/src/app/highlights/useHighlights.ts]:
```suggestion
  return useContextSelector(Context, selector);
```

File: packages/story-editor/src/app/highlights/provider.tsx [hunk 9]
```diff
@@ -17,52 +17,62 @@
...
+    showEffect?: boolean;
+    tab: string;
+    section: string;
+  };
+};
+
+function HighlightsProvider({ children }: PropsWithChildren<unknown>) {
+  const [highlighted, setHighlighted] = useState<
+    HighlightState | Record<never, never>
+  >({});
```
Comment (miina) [inline packages/story-editor/src/app/highlights/provider.tsx]: `Record<never, never>` is for the empty object? If you make `[k in HighlightType]:` above optional, it should work without it as well probably, no?

**Comment by miina** [comment]
The PR needs rebase + there are failing Karma tests.

File: packages/story-editor/tsconfig.json (hunk 4 above)
Comment (spacedmonkey) [reply packages/story-editor/tsconfig.json]: No style.js is not TS.

File: packages/story-editor/src/components/canvas/karma/quickActions.karma.js (hunk 7 above)
Comment (spacedmonkey) [reply packages/story-editor/src/components/canvas/karma/quickActions.karma.js:26]: This was an ask from @swissspidy

File: packages/story-editor/src/components/panels/design/animation/animation.js (hunk 6 above)
Comment (spacedmonkey) [reply packages/story-editor/src/components/panels/design/animation/animation.js]: Style.js not converted. So I had to side step this problem by loading it into a different import.

File: packages/story-editor/src/app/highlights/states.ts (hunk 5 above)
Comment (spacedmonkey) [reply packages/story-editor/src/app/highlights/states.ts]: I am not 100% sure where this should go either. This felt right, but willing to be talked out of it.

**Comment by spacedmonkey** [comment]
@miina Tests now passing.

File: packages/story-editor/src/components/canvas/karma/quickActions.karma.js (hunk 7 above)
Comment (swissspidy) [reply packages/story-editor/src/components/canvas/karma/quickActions.karma.js:26]: Ask is maybe the wrong word, but it definitely didn't feel right that `quickActions` is nested within `highlights`. It could easily be one level up, so I suggested that once. Makes it easier to separately convert `quickActions` later (see #12705)

Comment (miina) [reply packages/story-editor/src/components/canvas/karma/quickActions.karma.js:26]: True, it was quite confusing before.

File: packages/story-editor/src/app/highlights/states.ts (hunk 5 above)
Comment (miina) [reply packages/story-editor/src/app/highlights/states.ts]: Elsewhere we have left the enum into where the constants were before, for consistency, would be better to have it where before.

**Review by miina — APPROVED** [review]
LGTM pending the enum topic.

File: packages/story-editor/src/app/highlights/states.ts (hunk 5 above)
Comment (barklund) [reply packages/story-editor/src/app/highlights/states.ts]: _The Master of Enums_ henceforth declares, that we should keep enums in place of the constants they're replacing, because they (as @miina says) are transpiled to actual code and aren't stripped when type-checking. Also, this is what we've done elsewhere 🤷.

**Comment by spacedmonkey** [comment]
@miina With the enum disucssion resolved, we are we good to go?

With https://github.com/GoogleForCreators/web-stories-wp/pull/12891 merged, is the theme typed now? Can I remove this workaround this PR?

**Comment by miina** [comment]
> @miina With the enum disucssion resolved, we are we good to go?
> 
> With #12891 merged, is the theme typed now? Can I remove this workaround this PR?

Yep, good to go!

On theme -- it's typed only as much as I needed it in #12891, so I had to do[ this ](https://github.com/GoogleForCreators/web-stories-wp/pull/12891/files#diff-911e82ce531b8c9858211b1eb5a960afd11eb97419ecc138f8416b47e2303124R58)for example for it to work, doesn't work out of the box with TypeScript.

**Comment by spacedmonkey** [comment]
@miina Should we merge this then?

**Comment by miina** [comment]
> @miina Should we merge this then?

@spacedmonkey I can still see the enum in the `types` file though: https://github.com/GoogleForCreators/web-stories-wp/pull/12701/files#diff-0baf209296816c9ab521c82ebb3a938167eb7f6ce3bf0ec3dfaf4f0e8386de39R22

The enum should be just instead of where the `const` originally was.

**Comment by spacedmonkey** [comment]
> The enum should be just instead of where the `const` originally was.

@miina Sorry I missed this feedback. So just a convert these all back?

**Comment by miina** [comment]
> Sorry I missed this feedback. So just a convert these all back?

Just use `enum` instead of `const` without changing the location.

Initially it was 
```
const keys = {
  STORY_TITLE: 'STORY_TITLE',
  ELEMENT_TOOLBAR_TOGGLE: 'ELEMENT_TOOLBAR_TOGGLE',

  // Sidebar tabs
  STYLE_PANE: 'STYLE_PANE',
....
```

Instead of this, in the same file, you can just convert it to an enum:
```
 enum HighlightKeys
{
  StoryTitle = 'STORY_TITLE'
  ... etc.
}
```

This would still be in the `states.ts` file as before.

**Comment by spacedmonkey** [comment]
> This would still be in the `states.ts` file as before.

I don't see the reasoning for this. `HighlightType` is used in the type file. Why would I move it back to the state file and import it type file. This feels like the wrong way around. @miina

**Comment by miina** [comment]
> > This would still be in the `states.ts` file as before.
> 
> I don't see the reasoning for this. `HighlightType` is used in the type file. Why would I move it back to the state file and import it type file. This feels like the wrong way around. @miina

This is exactly the same conversion we had [here](https://github.com/GoogleForCreators/web-stories-wp/pull/12701#discussion_r1054403675).

That's how we have done it everywhere else as well. Since enum is not just a type but actually replaces the constant and holds values then we should keep it where it was before.

File: packages/story-editor/src/app/highlights/states.ts [hunk 10]
```diff
@@ -22,147 +22,141 @@ import { STYLE_PANE_IDS } from '@googleforcreators/design-system';
...
- * @typedef {Highlight} Highlight The current state of editor highlights
- * @property {string} tab The ID of the Sidebar tab to highlight
- * @property {boolean} focus Whether there is a focusable element, use for accessibility
- */
-
-const keys = {
-  STORY_TITLE: 'STORY_TITLE',
-  ELEMENT_TOOLBAR_TOGGLE: 'ELEMENT_TOOLBAR_TOGGLE',
+export enum HighlightType {
+  STORY_TITLE = 'STORY_TITLE',
```
Comment (miina) [inline packages/story-editor/src/app/highlights/states.ts]: The standard for enums is PascalCase (StoryTitle instead of `STORY_TITLE`), we've been following that in our codebase, too.

Comment (spacedmonkey) [reply packages/story-editor/src/app/highlights/states.ts]: Okay, it was a pain, but I converted all the enum to use PascalCase. This was 50 files changes and makes the PR much bigger.

File: packages/element-library/src/video/constants.js [hunk 11]
```diff
@@ -60,10 +60,10 @@ export const defaultAttributes = {
 };
 
 export const panels = [
-  PanelTypes.ELEMENT_ALIGNMENT,
+  PanelTypes.ElementAlignment,
```
Comment (miina) [inline packages/element-library/src/video/constants.js:63]: This is still constant though currently, not `enum`. 🤔

Comment (spacedmonkey) [reply packages/element-library/src/video/constants.js:63]: No it isn't. https://github.com/GoogleForCreators/web-stories-wp/pull/12701/files#diff-8277fbfe1e012d5dffa9b90466147be94743b1a9e409e5e1245e9ab20cf19614R17

**Review by miina — APPROVED** [review]
Looks like currently one constant is also converted to pascal case (either convert it to enum or keep it as it was). Otherwise looks good and ready to 🚢.

File: packages/element-library/src/video/constants.js (hunk 11 above)
Comment (miina) [reply packages/element-library/src/video/constants.js:63]: Oh, true, my bad :)

**Comment by swissspidy** [comment]
According to the Percy snapshots the "Border" panel is now in a different position than before. Used to be somewhere in the bottom (depending on element type) but now it's at the top:

Screenshot 2023-01-09 at 11 25 46

Screenshot 2023-01-09 at 11 26 12

Looks like `getDesignPanelsForSelection` grabs the values of `PanelTypes` as-is, so the order of entries there does matter.

Suggestion:

Change the order of items in `enum PanelTypes` to how it was before and add a comment that the order affects final position in the editor.

**Comment by spacedmonkey** [comment]
> Change the order of items in enum PanelTypes to how it was before and add a comment that the order affects final position in the editor.

Fixed in https://github.com/GoogleForCreators/web-stories-wp/commit/8c03c4126a16f098eec4c1cedbdc9323304a093d

**Review by swissspidy — APPROVED** [review]
